util/syspolicy/source/policy_store_windows.go

```diff
@@ -0,0 +1,454 @@
+	ps.locked.Done()
+}
+
+// RegisterChangeCallback adds a function that will be called whenever there's a policy change.
+// It returns a function that can be used to unregister the specified callback or an error.
+// The error is [ErrStoreClosed] if ps has already been closed.
+func (ps *PlatformPolicyStore) RegisterChangeCallback(cb func()) (unregister func(), err error) {
+	ps.mu.Lock()
+	defer ps.mu.Unlock()
+	if ps.closing {
+		return nil, ErrStoreClosed
+	}
+
+	handle := ps.cbs.Add(cb)
+	if len(ps.cbs) == 1 {
+		if ps.watcher, err = gp.NewChangeWatcher(ps.scope, ps.onChange); err != nil {
+			return nil, err
+		}
+	}
+
+	return func() {
+		ps.mu.Lock()
+		defer ps.mu.Unlock()
+		delete(ps.cbs, handle)
+		if len(ps.cbs) == 0 {
+			if ps.watcher != nil {
+				ps.watcher.Close()
+				ps.watcher = nil
+			}
+		}
+	}, nil
+}
+
+func (ps *PlatformPolicyStore) onChange() {
+	ps.mu.RLock()
+	defer ps.mu.RUnlock()
+	if ps.closing {
+		return
+	}
+	for _, callback := range ps.cbs {
+		go callback()
+	}
+}
+
+// ReadString retrieves a string policy with the specified key.
+// It returns [setting.ErrNotConfigured] if the policy setting does not exist.
+func (ps *PlatformPolicyStore) ReadString(key setting.Key) (val string, err error) {
+	return getPolicyValue(ps, key,
+		func(key registry.Key, valueName string) (string, error) {
+			val, _, err := key.GetStringValue(valueName)
+			return val, err
+		})
+}
+
+// ReadUInt64 retrieves an integer policy with the specified key.
+// It returns [setting.ErrNotConfigured] if the policy setting does not exist.
+func (ps *PlatformPolicyStore) ReadUInt64(key setting.Key) (uint64, error) {
+	return getPolicyValue(ps, key,
+		func(key registry.Key, valueName string) (uint64, error) {
+			val, _, err := key.GetIntegerValue(valueName)
+			return val, err
+		})
+}
+
+// ReadBoolean retrieves a boolean policy with the specified key.
+// It returns [setting.ErrNotConfigured] if the policy setting does not exist.
+func (ps *PlatformPolicyStore) ReadBoolean(key setting.Key) (bool, error) {
+	return getPolicyValue(ps, key,
+		func(key registry.Key, valueName string) (bool, error) {
+			val, _, err := key.GetIntegerValue(valueName)
+			if err != nil {
+				return false, err
+			}
+			return val != 0, nil
+		})
+}
+
+// ReadString retrieves a multi-string policy with the specified key.
+// It returns [setting.ErrNotConfigured] if the policy setting does not exist.
+func (ps *PlatformPolicyStore) ReadStringArray(key setting.Key) ([]string, error) {
+	return getPolicyValue(ps, key,
+		func(key registry.Key, valueName string) ([]string, error) {
+			val, _, err := key.GetStringsValue(valueName)
+			if err != registry.ErrNotExist {
+				return val, err // the err may be nil or non-nil
+			}
+
+			// The idiomatic way to store multiple string values in Group Policy
+			// and MDM for Windows is to have multiple REG_SZ (or REG_EXPAND_SZ)
+			// values under a subkey rather than in a single REG_MULTI_SZ value.
```

Comment:
Hmm, is that because `REG_MULTI_SZ` is a recent addition and there's been a need for such a thing for a long time? Just curious.

That's annoying but thanks for the attention to detail here.

Reply:
`REG_MULTI_SZ` was added back in the nineties. I'm not perfectly sure when exactly it became available, but it certainly pre-dates the Group Policy in general (added in NT 4.0), and ADMX specifically (added in Vista).

I'd say the main reasons why Group Policy / Administrative Templates don't use it for lists are:
- Mergeability (additivity?) of changes made in different GPOs applied to different OUs. That is, in certain cases if the same policy setting is configured in more than one place, the resultant "value" observed on the endpoints includes all the added items. And this doesn't require any special handling; the GP engine just writes values, one by one, as it reads the `Registry.pol` file from each applicable GPO.
- It's easier for to check policy settings this way, both in the user and kernel mode. Since the Name and the Value are the same for list items, an app or a driver could just check if a registry value with a given name exists if it needs to check if, say, a certain domain is okay to visit, is a member of a network zone, and so on. Otherwise, it would either have to re-read and check the entire list of values every time, or keep an in-memory copy of the most recent values as we're about to do. In most cases, it's much easier (and arguably faster in a lot of cases) to just check whether a registry value exists.
- It's compatible with another type of policy setting, which is essentially a string map. It's similar to the list setting type, but unlike the list, an admin can also configure a value for a given key. But it's the same code under the hood and requires no special handling on the admin console, GPClient service, or app side. If an app evolves, it can probably even transition from a list to a dictionary, and maintain compatibility with any existing policies.